[ios] Add permanent user heading indicator

friedbunny opened the pull request:
This is an WIP towards fixing #5523, adding a permanent user heading indicator (visible outside of heading tracking mode) and refactoring the user location annotation a bit.

## To-do
- [x] Handle beam ↔ arrow heading indicator switching.
- [x] Rotate the arrow properly.
- [x] Make `showsUserHeadingIndicator` inspectable.
- [x] Test that various tint colors are legible.
- [x] Document relationship between `showsUserHeadingIndicator` and `showsUserLocation`.
- [x] Reevaluate if c01c1ab0296b28b009e0f3d844dc42f40a6b67ee is necessary after #9845.
- [x] Make a call on whether or not to change the heading filter from 5° to 1°.
- [x] Clean up debug code.
- [x] Squash and reorganize commits.
- [x] Review from my lovely teammates.

/cc @1ec5 @fabian-guerra @jmkiley

friedbunny commented:
Preliminary design:

![img_5463](https://user-images.githubusercontent.com/1198851/29899048-152799a6-8db7-11e7-9a5c-8573f3b34af3.PNG)

Note sure if it would look better overlapping the white border or if it should stay outside. 🤔

fabian-guerra commented:
Is it possible see the contrast between the arrow heading <-> beam indicator with Mapbox Streets and Dark styles?

friedbunny commented:
e0b27fe633eb2ddf8a9c72adbdd75fae7c67bb23 adds a 1pt white stroke to the arrow and moves it on top of the outer white dot layer. I think this improves legibility with darker tint colors (and it’s kinda’ cute). 🐦 

![stroked-arrow](https://user-images.githubusercontent.com/1198851/29946988-90631e24-8e75-11e7-8ce4-a84a26fa0e4f.png)

Still need to test this with different pixel densities and tint colors.

1ec5 commented on platform/ios/src/MGLMapView.h:
```diff
@@ -358,6 +358,23 @@ MGL_EXPORT IB_DESIGNABLE
+
+ The default value of this property is `NO`.
+ */
+@property (nonatomic, assign) BOOL showsUserHeadingIndicator;
```
Should this property be inspectable? What happens if this property is set but `showsUserLocation` is not?

friedbunny replied:
> Should this property be inspectable?

Sure, that would be useful — I’ll do this.

> What happens if this property is set but `showsUserLocation` is not?

`showsUserHeadingIndicator` won’t do anything without `showsUserLocation=YES` — it doesn’t cause the user location annotation to appear. I considered having `showsUserHeadingIndicator` also implicitly enable `showsUserLocation` (the way our tracking modes do), but balked because it seemed a little too magical/indirect. Considering it again now, I could probably go either way. What do you think?

In any case, I’ll add documentation for this.

friedbunny replied:
Made this inspectable in 776a1c3 — renamed it for IB to try and disambiguate.

![screen shot 2017-09-01 at 3 11 08 pm](https://user-images.githubusercontent.com/1198851/29984342-04b950ba-8f28-11e7-895e-10240948846f.png)

friedbunny replied:
fa692f8: now with less truncation.

![screen shot 2017-09-01 at 3 24 11 pm](https://user-images.githubusercontent.com/1198851/29984933-c3d8a192-8f2a-11e7-9c1a-4364147da896.png)

friedbunny commented:
I’ve rebased, squashed, and broken this into digestible commits. I’m relatively happy with the state it’s in and would appreciate review from y’all. 🙇

A couple notes:
- The heading filter has been reduced to 1° (the default) from 5°, which smooths the rotation of the map and indicator. Performance (in terms of FPS) should be essentially the same as before, as the heading indicator no longer has implicit animations.
- Large heading changes will therefore cause the indicator to jump. Ideally we would animate large changes, but I’d prefer to leave that as tail work.

/cc @fabian-guerra @1ec5 @frederoni

1ec5 commented on platform/ios/src/MGLFaux3DUserLocationAnnotationView.h:
```diff
@@ -1,7 +1,12 @@
+const CGFloat MGLUserLocationAnnotationHaloSize = 115.0;
+
+const CGFloat MGLUserLocationAnnotationPuckSize = 45.0;
+const CGFloat MGLUserLocationAnnotationArrowSize = MGLUserLocationAnnotationPuckSize * 0.6;
```
Are these constants used anywhere outside of MGLFaux3DUserLocationAnnotationView.m? If not, I think they can stay in that file.

friedbunny replied:
They were during the initial refactor, but I ended up trying to avoid using these outside of `MGLFaux3DUserLocationAnnotationView.m`. I’ll see about moving these back to there.

friedbunny replied:
`MGLUserLocationHeadingBeamLayer` still needs `MGLUserLocationAnnotationHaloSize`, which unfortunately can’t be derived from the user location annotation itself.

1ec5 commented on platform/ios/src/MGLFaux3DUserLocationAnnotationView.m:
```diff
@@ -225,70 +215,67 @@ - (void)drawDot
+            CGFloat rotation = -MGLRadiansFromDegrees(self.mapView.direction - self.userLocation.heading.trueHeading);
+
+            // Don't rotate if the change is imperceptible (arbitrarily chosen as: 0.01 radians/~0.6°).
+            if (fabs(rotation) > 0.01)
```
Hoist this up top as a constant. Worth noting that this tolerance is much smaller than the other angular tolerance, `MGLToleranceForSnappingToNorth`, but I think that’s OK because the unrotation animation is meant to be perceptible.

friedbunny commented on platform/ios/app/MBXViewController.m:
```diff
@@ -1791,22 +1795,33 @@ - (void)mapView:(MGLMapView *)mapView didFinishLoadingStyle:(MGLStyle *)style
+        hudString = [NSString stringWithFormat:@"Visible: %ld  Queued: %ld", (unsigned long)self.mapView.visibleAnnotations.count, (unsigned long)queuedAnnotations];
     } else if (self.showZoomLevelEnabled) {
-        self.hudLabel.text = [NSString stringWithFormat:@" Zoom: %.2f", self.mapView.zoomLevel];
+        hudString = [NSString stringWithFormat:@"%.2f ∕ ↕\U0000FE0E%.f° ∕ %.f°", self.mapView.zoomLevel, self.mapView.camera.pitch, self.mapView.direction];
```
BTW, this ends up looking like this in the label:
> 14.12  ∕ ↕&#xfe0e;45° ∕ 280°

... thanks to the good ole [variation selector](http://www.fileformat.info/info/unicode/char/fe0e/index.htm), which negates emoji conversion.

friedbunny commented:
Tests are failing on the release branch because of `egl-node6-clang39-debug` builds on Travis, [such as this](https://travis-ci.org/mapbox/mapbox-gl-native/jobs/273113609). Are you aware of any changes in master that we can cherry-pick to fix this, @brunoabinader @bsudekum?

brunoabinader commented:
Not really @friedbunny :disappointed: this is the first time I see this failure.
